Fix #243: Use of Python "with" semantic in API client to guarantee speculos instance termination

kingofpayne opened the pull request.

greenknot commented on speculos/client.py:
```diff
@@ -185,15 +185,19 @@ def stop(self) -> None:
+        Api.__init__(self, self.__api_url)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
```
Please add type annotation for all arguments.

FYI, it was found with:

```console
$ mypy --strict speculos/client.py
speculos/client.py:199: error: Function is missing a type annotation for one or more arguments
```

unfortunately `mypy` warnings should be fixed across the entire code base before being integrated to the CI.

greenknot reviewed (changes requested):
Thanks @kingofpayne. I'm not familiar with https://docs.python.org/3/library/contextlib.html, however it LGTM.

kingofpayne commented:
Updated `client.py` to start and stop correctly both speculos and event streaming.
It is also now possible to call manually `start` and `stop` on `SpeculosInstance` if the user is not willing to use `with` statement.

Also added typing for `SpeculosInstance.__exit__`.
